README.md
# @allex/logger

A light logger extends the builtin console module with some log enhance.

## Installation

```bash
$ npm i @allex/logger
```

## Usage

```js
import { getLogger } from '@allex/logger'

let logger = getLogger()

logger.setLevel('debug') // error, warn, info, log, debug, silent

logger.log('Start to logging... (logLevel: debug)')
logger.debug('I\'m just a debug info, you can skip it.')
logger.log('JSON', { foo: 1, arr: [1, 2, 3] })
logger.info('Hi, i\'m a ammmm info.')
logger.warn('Warning, (pay) attention please.')
logger.error('Fatal, sorry, some thing fails.')

console.log('\n----------------- create a new logger instance -----------------\n')

logger = getLogger({
  timeStamp: true,
  logLevel: 'info'
})

logger.log('Start to logging... (logLevel: info)')
logger.debug('I\'m just a debug info, you can skip it.')
logger.info('Hi, i\'m a ammmm info.')
logger.warn('Waring, (pay) attention please.')
logger.error('Fatal, sorry, some thing fails.')
```

## APIs

default options:

```js
const defaultOptions: LoggerOptions = {
  colour: true,
  timeStamp: false,
  prefix: true,
  logLevel: 'info'
}
```

```ts
export declare type LogLevel = 'debug' | 'log' | 'info' | 'warn' | 'error' | 'silent';
interface IBaseLogger {
    log(...args: any[]): void;
    debug(...args: any[]): void;
    info(...args: any[]): void;
    warn(...args: any[]): void;
    error(...args: any[]): void;
}
export interface ILogger extends IBaseLogger {
    setLevel(level: LogLevel): void;
    setProvider(ctor: () => IBaseLogger): void;
}
export interface LoggerOptions {
    colour: boolean;
    timeStamp: boolean;
    prefix: boolean;
    logLevel: LogLevel;
}
export declare const getLogger: (options: Partial<LoggerOptions>) => ILogger;
```

## License

MIT


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

### [2.0.1](https://github.com/allex/logger/compare/2.0.0...2.0.1) (2021-04-12)


### Bug Fixes

* getLogger allow none options ([679f83f](https://github.com/allex/logger/commit/679f83f80ba9a1ea0709361953364898e159a5a2))
* set timeStamp to false by the default ([a0416b1](https://github.com/allex/logger/commit/a0416b140cb1c38c22be5b92cf011f4bb494be08))

## [2.0.0](https://github.com/allex/logger/compare/1.0.2...2.0.0) (2021-04-12)


### Features

* refactor logger factory with multiple instance, transform tsc ([5dbdbe8](https://github.com/allex/logger/commit/5dbdbe8f3876845b66f8ea17f261233b6266c680))


test/test.js
const { getLogger } = require('../')

let logger = getLogger()

logger.setLevel('debug') // error, warn, info, log, debug, silent

logger.log('Start to logging... (logLevel: debug)')
logger.debug('I\'m just a debug info, you can skip it.')
logger.log('JSON', { foo: 1, arr: [1, 2, 3] })
logger.info('Hi, i\'m a ammmm info.')
logger.warn('Warning, (pay) attention please.')
logger.error('Fatal, sorry, some thing fails.')

console.log('\n----------------- create a new logger instance -----------------\n')

logger = require('..').getLogger({
  timeStamp: true,
  logLevel: 'info'
})

logger.log('Start to logging... (logLevel: info)')
logger.debug('I\'m just a debug info, you can skip it.')
logger.info('Hi, i\'m a ammmm info.')
logger.warn('Waring, (pay) attention please.')
logger.error('Fatal, sorry, some thing fails.')


.fssrc.js
// vim: set ft=javascript fdm=marker et ff=unix tw=80 sw=2:

import path from 'path'

const {
  name,
  version,
  license,
  author,
  dependencies,
  repository
} = require('./package.json')

const banner = (name, short = false) => {
  let s
  if (short) {
    s = `/*! ${name} v${version} | ${license} licensed | ${author} */`
  } else {
    s = `/**
 * ${name} v${version} - ${repository}
 *
 * @author ${author}
 * Released under the ${license} license.
 */`
  }
  return s
}

const plugins = [
  'resolve',
  'typescript',
  'babel'
]

export default {
  destDir: path.join(__dirname, './lib'),
  dependencies,
  entry: [
    {
      input: 'src/index.ts',
      plugins,
      targets: [
        { format: 'umd', name: 'Logger', file: 'logger.min.js', banner: banner(name) },
        { format: 'cjs', file: 'logger.cjs.js', banner: banner(name) },
        { format: 'es', file: 'logger.es.js', banner: banner(name) }
      ]
    }
  ]
}


src/index.js
/**
 * A util extends the built-in console module with some log enhances.
 *
 * @author Allex Wang <allex.wxn@gmail.com>
 * Released under the MIT License.
 */

import sprintf from 'sprintf-js'
import leftPad from './leftPad'
import styles from './styles'

const pad = (s, n, c) => {
  n = n || 2
  c = c === undefined ? '0' : c
  return leftPad(s, n, c)
}

// 26 Feb 16:19:34
const ts = () => {
  const d = new Date()
  const padn = n => pad(n, 2)
  const date = [d.getFullYear(), d.getMonth() + 1, d.getDate()].map(padn).join('-')
  let time = [
    d.getHours(),
    d.getMinutes(),
    d.getSeconds()
  ].map(padn).join(':')
  time += '.' + pad(d.getMilliseconds(), 3)
  return [ date, time ].join(' ')
}

const LOG_TYPES = {
  error: [ 0, 'red', 'x' ],
  warn: [ 1, 'yellow', '!' ],
  info: [ 2, 'cyan', 'i' ],
  log: [ 3, '', '' ],
  debug: [ 4, '', '*' ]
}

const getLogPrefix = type => LOG_TYPES[type][2] || ''

// Returns true if running in interactive shell
export const isTTY = !!(process.stderr || 0).isTTY

// Returns a colorize string by specific color type
// colorize(str, type);
export const colorize =
  isTTY ? (s, type) => {
    let style = type && styles[type]
    return style ? style[0] + s + style[1] : s
  } : (s, type) => s

export class Logger {
  constructor () {
    this.level = process.env.IMLOG_LEVEL || 4
    let logger = global.console

    // Provide logger utilities with colors
    ;['log', 'info', 'error', 'warn', 'debug'].forEach(k => {
      var type = LOG_TYPES[k][0]
      this[k] = function (s, ...args) {
        if (this.level < type) { return }

        if (s && typeof s === 'object') {
          s = JSON.stringify(s)
        }

        // log prefix
        var prefix = getLogPrefix(k)
        if (prefix) {
          s = '[' + prefix + '] ' + s
        }

        // timestamp
        s = '[' + ts() + '] ' + s

        if (isTTY) {
          s = colorize(s, LOG_TYPES[k][1])
        }

        return logger[k](s, ...args)
      }.bind(this)
    })

    this.setLevel = function (level) {
      this.level = level
      // write to env var to make logger level cross subshell ability
      process.env.IMLOG_LEVEL = level
    }.bind(this)
  }
}

const defaultLogger = new Logger()

export const { setLevel, error, warn, info, log, debug } = defaultLogger

// out print test to stderr
export const stderr = s => {
  process.stderr.write(colorize(s, LOG_TYPES.error[1]))
}

// out print text to stdout
export const stdout = s => {
  process.stdout.write(s)
}

export const print = (...args) => {
  const s = sprintf(...args)
  process.stdout.write(s)
}

export const puts = (...args) => {
  const s = sprintf(...args)
  process.stdout.write(s + '\n')
}

// Return a formatted string, util.format enhancements
export { sprintf }


src/styles.js
export default {
  // styles
  'bold': ['\x1B[1m', '\x1B[22m'],
  'italic': ['\x1B[3m', '\x1B[23m'],
  'underline': ['\x1B[4m', '\x1B[24m'],
  'inverse': ['\x1B[7m', '\x1B[27m'],
  'strikethrough': ['\x1B[9m', '\x1B[29m'],
  // text colors
  // grayscale
  'white': ['\x1B[37m', '\x1B[39m'],
  'grey': ['\x1B[90m', '\x1B[39m'],
  'black': ['\x1B[30m', '\x1B[39m'],
  // colors
  'blue': ['\x1B[34m', '\x1B[39m'],
  'cyan': ['\x1B[36m', '\x1B[39m'],
  'green': ['\x1B[32m', '\x1B[39m'],
  'magenta': ['\x1B[35m', '\x1B[39m'],
  'red': ['\x1B[31m', '\x1B[39m'],
  'yellow': ['\x1B[33m', '\x1B[39m'],
  // background colors
  // grayscale
  'whiteBG': ['\x1B[47m', '\x1B[49m'],
  'greyBG': ['\x1B[49;5;8m', '\x1B[49m'],
  'blackBG': ['\x1B[40m', '\x1B[49m'],
  // colors
  'blueBG': ['\x1B[44m', '\x1B[49m'],
  'cyanBG': ['\x1B[46m', '\x1B[49m'],
  'greenBG': ['\x1B[42m', '\x1B[49m'],
  'magentaBG': ['\x1B[45m', '\x1B[49m'],
  'redBG': ['\x1B[41m', '\x1B[49m'],
  'yellowBG': ['\x1B[43m', '\x1B[49m']
}


src/styles.ts
export default {
  // styles
  'bold': ['\x1B[1m', '\x1B[22m'],
  'italic': ['\x1B[3m', '\x1B[23m'],
  'underline': ['\x1B[4m', '\x1B[24m'],
  'inverse': ['\x1B[7m', '\x1B[27m'],
  'strikethrough': ['\x1B[9m', '\x1B[29m'],

  // text colors
  // grayscale
  'white': ['\x1B[37m', '\x1B[39m'],
  'grey': ['\x1B[90m', '\x1B[39m'],
  'black': ['\x1B[30m', '\x1B[39m'],

  // colors
  'blue': ['\x1B[34m', '\x1B[39m'],
  'cyan': ['\x1B[36m', '\x1B[39m'],
  'green': ['\x1B[32m', '\x1B[39m'],
  'magenta': ['\x1B[35m', '\x1B[39m'],
  'red': ['\x1B[31m', '\x1B[39m'],
  'yellow': ['\x1B[33m', '\x1B[39m'],

  // background colors
  // grayscale
  'whiteBG': ['\x1B[47m', '\x1B[49m'],
  'greyBG': ['\x1B[49;5;8m', '\x1B[49m'],
  'blackBG': ['\x1B[40m', '\x1B[49m'],

  // colors
  'blueBG': ['\x1B[44m', '\x1B[49m'],
  'cyanBG': ['\x1B[46m', '\x1B[49m'],
  'greenBG': ['\x1B[42m', '\x1B[49m'],
  'magentaBG': ['\x1B[45m', '\x1B[49m'],
  'redBG': ['\x1B[41m', '\x1B[49m'],
  'yellowBG': ['\x1B[43m', '\x1B[49m']
}


src/leftPad.js
/* This program is free software. It comes without any warranty, to
     * the extent permitted by applicable law. You can redistribute it
     * and/or modify it under the terms of the Do What The Fuck You Want
     * To Public License, Version 2, as published by Sam Hocevar. See
     * http://www.wtfpl.net/ for more details. */
'use strict'
module.exports = leftPad

var cache = [
  '',
  ' ',
  '  ',
  '   ',
  '    ',
  '     ',
  '      ',
  '       ',
  '        ',
  '         '
]

function leftPad (str, len, ch) {
  // convert `str` to a `string`
  str = str + ''
  // `len` is the `pad`'s length now
  len = len - str.length
  // doesn't need to pad
  if (len <= 0) return str
  // `ch` defaults to `' '`
  if (!ch && ch !== 0) ch = ' '
  // convert `ch` to a `string` cuz it could be a number
  ch = ch + ''
  // cache common use cases
  if (ch === ' ' && len < 10) return cache[len] + str
  // `pad` starts with an empty string
  var pad = ''
  // loop
  while (true) {
    // add `ch` to `pad` if `len` is odd
    if (len & 1) pad += ch
    // divide `len` by 2, ditch the remainder
    len >>= 1
    // "double" the `ch` so this operation count grows logarithmically on `len`
    // each time `ch` is "doubled", the `len` would need to be "doubled" too
    // similar to finding a value in binary search tree, hence O(log(n))
    if (len) ch += ch
    // `len` is 0, exit the loop
    else break
  }
  // pad `str`!
  return pad + str
}


src/index.ts
/**
 * A util extends the built-in console module with some log enhances.
 *
 * @author Allex Wang <allex.wxn@gmail.com>
 * Released under the MIT License.
 */

import util from 'util'

import styles from './styles'

export type LogLevel = 'debug' | 'log' | 'info' | 'warn' | 'error' | 'silent'

interface IBaseLogger {
  log(...args: any[]): void;
  debug(...args: any[]): void;
  info(...args: any[]): void;
  warn(...args: any[]): void;
  error(...args: any[]): void;
}

export interface ILogger extends IBaseLogger {
  setLevel(level: LogLevel): void;
  setProvider(ctor: () => IBaseLogger): void;
}

export interface LoggerOptions {
  colour: boolean;
  timeStamp: boolean;
  prefix: boolean;
  logLevel: LogLevel;
}

type Transformer = <T = any>(...args: any[]) => T;

// cache instance by params
let loggerInstances = Object.create(null)

const compose = (...fns: Transformer[]) => {
  if (fns.length === 0) {
    throw new Error('compose requires at least one argument')
  }

  return (...args: any[]) => {
    const list = fns.slice()
    if (list.length > 0) {
      let result
      while (list.length > 0) {
        result = list.pop()(...args)
        args[0] = result
      }
      return result
    }
  }
}

const getPadSpace = (() => {
  const cache = Object.create(null)
  return (c: string, n: number): string => {
    const k = `${c}:${n}`
    return cache[k] || (cache[k] = new Array(n).fill(c).join(''))
  }
})()

const pad = (s: number, n: number = 2) => (getPadSpace('0', n) + s).substr(-1 * n)

// 2020-04-02 16:19:34.382
const ts = () => {
  const d = new Date()
  const date = [d.getFullYear(), d.getMonth() + 1, d.getDate()].map(pad).join('-')
  let time = [d.getHours(), d.getMinutes(), d.getSeconds()].map(pad).join(':')
  time += '.' + pad(d.getMilliseconds(), 3)
  return [ date, time ].join(' ')
}

// Returns true if running in interactive shell
const isTTY = !!(process.stderr || {}).isTTY

const c = console
const defaultProvider: IBaseLogger = {
  log: c.log,
  debug: c.log, // use .log(); since console does not have .debug()
  info: c.info,
  warn: c.warn,
  error: c.error
}

// log level 'weight'
const LEVELS = {
  debug: 1,
  info: 2,
  warn: 3,
  error: 5,
  silent: 8
}

const LEVEL_CFG = {
  error: [ 'ERROR', 'red' ],
  warn: [ 'WARN', 'yellow' ],
  info: [ 'INFO', 'cyan' ],
  log: [ '', '' ],
  debug: [ 'DEBUG', '' ]
}

type LogDecorator = (msg: string, level: LogLevel) => string

// Returns a colorize string by specific log level
// colorize(message, level);
const colorize: LogDecorator = isTTY
  ? (msg, level) => {
    let style = level && styles[LEVEL_CFG[level][1]]
    return style
      ? style[0] + msg + style[1]
      : msg
  }
  : (msg, level) => msg

const prefix: LogDecorator = (msg, level) => {
  const cfg = LEVEL_CFG[level]
  const t = cfg && cfg[0] || ''
  return t ? `[${t}] ${msg}` : msg
}

const timeStamp: LogDecorator = (msg, level) => `${ts()} - ${msg}`

const defaultOptions: LoggerOptions = {
  colour: true,
  timeStamp: false,
  prefix: true,
  logLevel: 'info'
}

function Logger (options: Partial<LoggerOptions>): ILogger {
  let logLevel: LogLevel
  let provider: IBaseLogger
  let fPrint: LogDecorator = s => s

  const api: ILogger = {
    log,
    debug,
    info,
    warn,
    error,
    setLevel,
    setProvider
  }

  init({ ...defaultOptions, ...options })

  return api

  function init (options: LoggerOptions): void {
    api.setLevel(options.logLevel || 'warn')
    api.setProvider(() => defaultProvider)

    const decorators = []
    if (options.colour) {
      decorators.push(colorize)
    }
    if (options.timeStamp) {
      decorators.push(timeStamp)
    }
    decorators.push(prefix)
    fPrint = compose(...decorators)
  }

  function setLevel (v: LogLevel): void {
    const isValid = Object.keys(LEVELS).includes(v)
    if (!isValid) {
      throw new Error(`Log level error. Invalid logLevel.`)
    }
    logLevel = v
  }

  function setProvider (ctor: () => IBaseLogger): void {
    if (!ctor || typeof ctor !== 'function') {
      throw new Error(`Log provider config error. Expecting a function.`)
    }
    provider = ctor()
  }

  // log will log messages, regardless of logLevels
  function log () { provider.log(_interpolate('log', arguments)) }
  function debug () { _showLevel('debug') && provider.debug(_interpolate('debug', arguments)) }
  function info () { _showLevel('info') && provider.info(_interpolate('info', arguments)) }
  function warn () { _showLevel('warn') && provider.warn(_interpolate('warn', arguments)) }
  function error () { _showLevel('error') && provider.error(_interpolate('error', arguments)) }

  /**
   * Decide to log or not to log, based on the log levels 'weight'
   * @param  {String} showLevel [debug, info, warn, error, silent]
   * @return {Boolean}
   */
  function _showLevel (showLevel: LogLevel): boolean {
    const currentLogLevel = LEVELS[logLevel]
    return currentLogLevel && currentLogLevel <= LEVELS[showLevel]
  }

  // make sure logged messages and its data are return interpolated
  // make it possible for additional log data, such date/time or custom prefix.
  function _interpolate (level: LogLevel, args: IArguments): string {
    const result = util.format(...args)
    return fPrint(result, level)
  }
}

// singleton
export const getLogger = (options?: Partial<LoggerOptions>): ILogger => {
  options = { ...defaultOptions, ...options }

  // cache instance by options
  const k = Object.keys(options).map(k => `${k}:${options[k]}`).join(',')

  if (!loggerInstances[k]) {
    loggerInstances[k] = Logger(options)
  }

  return loggerInstances[k]
}
